streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/ClientTagAwareStandbyTaskAssignor.java

```diff
@@ -0,0 +1,203 @@
+                                                 final Map<String, Set<UUID>> tagValueToClients,
+                                                 final Map<String, Set<String>> tagKeyToTagValues) {
+        for (final Entry<UUID, ClientState> clientStateEntry : clientStates.entrySet()) {
+            final UUID clientId = clientStateEntry.getKey();
+            final ClientState clientState = clientStateEntry.getValue();
+
+            clientState.clientTags().forEach((tagKey, tagValue) -> {
+                tagKeyToTagValues.computeIfAbsent(tagKey, ignored -> new HashSet<>()).add(tagValue);
+                tagValueToClients.computeIfAbsent(tagValue, ignored -> new HashSet<>()).add(clientId);
+            });
+        }
+    }
+
+    private static void assignStandbyTasksForActiveTask(final int numStandbyReplicas,
+                                                        final TaskId activeTaskId,
+                                                        final UUID activeTaskClient,
+                                                        final Set<String> rackAwareAssignmentTags,
+                                                        final Map<UUID, ClientState> clientStates,
+                                                        final Map<TaskId, Integer> tasksToRemainingStandbys,
+                                                        final Map<String, Set<String>> tagKeyToTagValues,
+                                                        final Map<String, Set<UUID>> tagValueToClients) {
+        final ConstrainedPrioritySet standbyTaskClientsByTaskLoad = new ConstrainedPrioritySet(
+            (client, t) -> !clientStates.get(client).hasAssignedTask(t),
+            client -> clientStates.get(client).assignedTaskLoad()
+        );
+
+        final Set<UUID> usedClients = new HashSet<>();
+
+        standbyTaskClientsByTaskLoad.offerAll(clientStates.keySet());
+
+        int numRemainingStandbys = tasksToRemainingStandbys.get(activeTaskId);
+
+        usedClients.add(activeTaskClient);
+
+        while (numRemainingStandbys > 0) {
+            final Set<UUID> clientsOnAlreadyUsedTagDimensions = findClientsOnUsedTagDimensions(
+                usedClients,
+                rackAwareAssignmentTags,
+                clientStates,
+                tagValueToClients,
+                tagKeyToTagValues
+            );
+
+            final UUID polledClient = standbyTaskClientsByTaskLoad.poll(
+                activeTaskId, uuid -> !clientsOnAlreadyUsedTagDimensions.contains(uuid)
+            );
+
+            if (polledClient == null) {
+                break;
+            }
+
+            final ClientState standbyTaskClient = clientStates.get(polledClient);
+
+            standbyTaskClient.assignStandby(activeTaskId);
+
+            usedClients.add(polledClient);
+
+            standbyTaskClientsByTaskLoad.offerAll(clientsOnAlreadyUsedTagDimensions);
```

Comment:
I do not understand why you re-add `clientsOnAlreadyUsedTagDimensions`. Those clients were not modified and not polled for sure due to line 140.

Reply:
We need to add those clients back because we need to fallback to least loaded clients if we don't have any more valid clients on other dimensions.
For instance, without this `shouldDistributeTasksOnLeastLoadedClientsWhenThereAreNoEnoughUniqueTagDimensions` test will fail.

Reply:
Are you sure, because I cannot confirm the failure of the test on my side?

Reply:
Yeah, sorry. You're right. This is not needed.

Reply:
I've removed this line and pushed the changes.